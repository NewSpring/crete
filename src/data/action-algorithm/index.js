import { ActionAlgorithm } from '@apollosproject/data-connector-rock';
import ApollosConfig from '@apollosproject/config';
import { get, flatten } from 'lodash';

const { schema, resolver } = ActionAlgorithm;

class dataSource extends ActionAlgorithm.dataSource {
  ACTION_ALGORITHMS = {
    ...this.ACTION_ALGORITHMS,
    CAMPAIGN_ITEMS: this.campaignItemsAlgorithm.bind(this),
    STAFF_NEWS: this.contentChannelAlgorithm.bind(this),
  };

  async campaignItemsAlgorithm({ limit = 1 } = {}) {
    const { ContentItem } = this.context.dataSources;

    const channels = await ContentItem.byContentChannelIds(
      ApollosConfig.ROCK_MAPPINGS.CAMPAIGN_CHANNEL_IDS
    ).get();

    const items = flatten(
      await Promise.all(
        channels.map(async ({ id, title }) => {
          const childItemsCursor = await ContentItem.getCursorByParentContentItemId(
            id
          );

          const childItems = await childItemsCursor
            .orderBy('StartDateTime', 'desc') // Custom for newspring. We are ordering by startDateTime here.
            .top(limit)
            .expand('ContentChannel')
            .get();

          return childItems.map((item) => ({
            ...item,
            channelSubtitle: title,
          }));
        })
      )
    );

    return items.map((item, i) => ({
      id: `${item.id}${i}`,
      title: item.title,
      subtitle: get(item, 'contentChannel.name'),
      relatedNode: { ...item, __type: ContentItem.resolveType(item) },
      image: ContentItem.getCoverImage(item),
      action: 'READ_CONTENT',
      summary: ContentItem.createSummary(item),
    }));
  }
}

export { dataSource, schema, resolver };
