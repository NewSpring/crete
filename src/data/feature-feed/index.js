import { createGlobalId, resolverMerge } from '@apollosproject/server-core';
import ApollosConfig from '@apollosproject/config';
import { FeatureFeed } from '@apollosproject/data-connector-rock';

const resolverChanges = {
  Query: {
    homeFeedFeatures: (root, args, { dataSources }, info) => {
      info.cacheControl.setCacheHint({ maxAge: 0 });
      return dataSources.FeatureFeed.getFeed({
        type: 'apollosConfig',
        args: { section: 'HOME_FEATURES', ...args },
      });
    },
  },
};

const resolver = resolverMerge(resolverChanges, FeatureFeed);
class dataSource extends FeatureFeed.dataSource {
  getFromId = (id) => this.getFeed(JSON.parse(id));

  getFilteredFeatures = async (args) => {
    const { Person, Auth, Group } = this.context.dataSources;
    const config = ApollosConfig[args.section] || [];

    const { id } = await Auth.getCurrentPerson();
    const isStaff = await Person.isStaff(id);
    const testGroups = await Group.getTestGroups(id);

    return config.filter((feature) => {
      // filter staff only features
      const isStaffFeature = feature.algorithms.filter(
        ({ type }) => type === 'STAFF_NEWS'
      ).length;
      if (isStaffFeature && !isStaff) return false;
      // filter experimental features
      const isExperimentalFeature = feature.algorithms.filter(
        ({ type }) => type === 'UPCOMING_EVENTS'
      ).length;
      const isTester = testGroups.filter(
        (group) => group.name === 'Experimental Features'
      ).length;
      if (isExperimentalFeature && !isTester) return false;
      return true;
    });
  };

  getFeed = async ({ type = '', args = {} }) => {
    let getFeatures = () => [];
    const { ContentItem, Feature } = this.context.dataSources;

    if (type === 'apollosConfig') {
      getFeatures = async () => {
        const features = await this.getFilteredFeatures(args);
        return Feature.getFeatures(features, args);
      };
    }

    if (type === 'contentItem' && args.id) {
      const contentItem = await ContentItem.getFromId(args.id);
      getFeatures = () => ContentItem.getFeatures(contentItem);
    }

    return {
      __typename: 'FeatureFeed',
      id: createGlobalId(JSON.stringify({ type, args }), 'FeatureFeed'),
      getFeatures,
    };
  };
}

export { resolver, dataSource };
