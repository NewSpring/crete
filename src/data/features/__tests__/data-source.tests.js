import ApollosConfig from '@apollosproject/config';
import { dataSource as FeatureDataSource } from '../index';
import { dataSource as FeedDataSource } from '../../feature-feed';
import { dataSource as ActionAlgorithm } from '../../action-algorithm';

async function expandResult(result) {
  let expandedResult = { ...result };
  if (result.cards && typeof result.cards === 'function') {
    expandedResult = { ...expandedResult, cards: await result.cards() };
  }
  if (result.actions && typeof result.actions === 'function') {
    expandedResult = { ...expandedResult, actions: await result.actions() };
  }
  if (result.heroCard && typeof result.heroCard === 'function') {
    expandedResult = { ...expandedResult, heroCard: await result.heroCard() };
  }
  if (result.prayers && typeof result.prayers === 'function') {
    expandedResult = { ...expandedResult, prayers: await result.prayers() };
  }

  return expandedResult;
}

ApollosConfig.loadJs({
  HOME_FEATURES: [
    {
      title: 'Staff news',
      algorithms: [{ type: 'STAFF_NEWS' }],
      type: 'ActionList',
    },
    {
      title: 'Events',
      algorithms: [{ type: 'UPCOMING_EVENTS' }],
      type: 'ActionList',
    },
  ],
});

describe('Feature data sources', () => {
  let Feature;
  let FeatureFeed;
  const getCursorByParentContentItemId = jest.fn(() => ({
    orderBy: () => ({
      top: () => ({
        expand: () => ({
          get: () => [
            {
              id: 1,
              title: 'first item',
              contentChannel: { name: 'content channel' },
            },
            {
              id: 2,
              title: 'second item',
              contentChannel: { name: 'content channel' },
            },
            {
              id: 3,
              title: 'third item',
              contentChannel: { name: 'content channel' },
            },
          ],
        }),
      }),
    }),
  }));
  const byContentChannelIds = jest.fn(() => ({
    get: () => [
      {
        id: 1,
        title: 'first item',
        contentChannel: { name: 'content channel' },
      },
      {
        id: 2,
        title: 'second item',
        contentChannel: { name: 'content channel' },
      },
      {
        id: 3,
        title: 'third item',
        contentChannel: { name: 'content channel' },
      },
    ],
  }));
  let context;
  beforeEach(() => {
    const ActionAlgo = new ActionAlgorithm();
    Feature = new FeatureDataSource();
    FeatureFeed = new FeedDataSource();
    context = { dataSources: {} };
    context.dataSources = {
      Auth: { getCurrentPerson: () => ({ id: 1 }) },
      ActionAlgorithm: ActionAlgo,
      Feature,
      Person: { isStaff: () => true },
      Group: {
        getTestGroups: () => [{ id: 1, name: 'Experimental Features' }],
      },
      FeatureFeed,
      ContentItem: {
        getCursorByParentContentItemId,
        byContentChannelIds,
        resolveType: () => 'UniversalContentItem',
        getCoverImage: () => null,
        createSummary: () => 'some summary',
      },
    };
    context.dataSources.ActionAlgorithm.initialize({ context });
    context.dataSources.Feature.initialize({ context });
    context.dataSources.FeatureFeed.initialize({ context });
  });
  it('gets available features', async () => {
    expect(await Feature.getHomeFeedFeatures()).toMatchSnapshot();
  });
  it('gets available as non-staff', async () => {
    FeatureFeed.context.dataSources.Person.isStaff = () => false;
    expect(await Feature.getHomeFeedFeatures()).toMatchSnapshot();
  });
  it('should create an ActionListFeature from a CAMPAIGN_ITEMS algorithm', async () => {
    const result = await Feature.createActionListFeature({
      algorithms: [
        {
          type: 'CAMPAIGN_ITEMS',
        },
      ],
      title: 'Test Featured Item',
      subtitle: "It's featured!",
    });

    expect(await expandResult(result)).toMatchSnapshot();
    expect(
      context.dataSources.ContentItem.getCursorByParentContentItemId.mock.calls
    ).toMatchSnapshot();
    expect(
      context.dataSources.ContentItem.byContentChannelIds.mock.calls
    ).toMatchSnapshot();
  });
});
