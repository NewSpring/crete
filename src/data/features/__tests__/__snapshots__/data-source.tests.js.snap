// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature data sources gets available as non-staff 1`] = `
Array [
  Object {
    "__typename": "ActionListFeature",
    "actions": [Function],
    "id": "{\\"algorithms\\":[{\\"type\\":\\"UPCOMING_EVENTS\\"}],\\"title\\":\\"Events\\"}",
    "primaryAction": undefined,
    "subtitle": undefined,
    "title": "Events",
  },
]
`;

exports[`Feature data sources gets available features 1`] = `
Array [
  Object {
    "__typename": "ActionListFeature",
    "actions": [Function],
    "id": "{\\"algorithms\\":[{\\"type\\":\\"STAFF_NEWS\\"}],\\"title\\":\\"Staff news\\"}",
    "primaryAction": undefined,
    "subtitle": undefined,
    "title": "Staff news",
  },
  Object {
    "__typename": "ActionListFeature",
    "actions": [Function],
    "id": "{\\"algorithms\\":[{\\"type\\":\\"UPCOMING_EVENTS\\"}],\\"title\\":\\"Events\\"}",
    "primaryAction": undefined,
    "subtitle": undefined,
    "title": "Events",
  },
]
`;
